lib/OptimizeMiniRust.ml

```diff
@@ -315,6 +331,21 @@ let rec infer_expr (env: env) valuation (return_expected: typ) (expected: typ) (
         failwith "TODO: operator not supported"
     end
 
+  | Call (Field (_, f, st) as fld, [], args) ->
+    let fields = Hashtbl.find structs (`Struct (assert_name st)) in
+    let field = List.find (fun fld -> fld.name = f) fields in
+    begin match field.typ with
+    | Function (_, arg_tys, _) ->
+      let known, args = List.fold_left2 (fun (known, args) arg arg_ty ->
+          let known, arg = infer_expr env valuation return_expected arg_ty known arg in
+          known, arg :: args
+        ) (known, []) args arg_tys
+      in
```

Comment:
aren't you missing a `List.rev args` here?

Reply:
You're totally right, that's why I usually stay far away from folds...

Reply:
What do you typically use then?

Reply:
In cases like this, map + mutable reference/state monad.

Using combinators like map and filters has the obvious advantage that you can't screw up the order or forget elements, and you don't need to do the accumulator yourself.  Reasoning about filter/map is easier too, since it's just first order and you don't need to figure out loop invariants.